TELCODOCS-597:  Adding a local registry to a Single Node OpenShift DU profile (#50904)
openshift/openshift-docs: modules/ztp-add-local-reg-for-sno-duprofile.adoc

// Module included in the following assemblies:
//
// scalability_and_performance/ztp-deploying-disconnected.adoc

:_module-type: CONCEPT
[id="ztp-add-local-reg-for-sno-duprofile_{context}"]
= Configuring the Image Registry Operator for local caching of images

{Product} manages image caching with a local registry. A Single Node OpenShift (SNO) Distributed Unit (DU) is subject to bandwidth restrictions when communicating with the centralized registry. This can lead to long image download times.

This is unavoidable for the initial deployment. Over time, there is a risk that `crio` will wipe `/var/lib/containers/storage` in the case of a disorderly shutdown.

> To address this, you need a local registry on the SNO. The cluster images need to be  mirrored to this registry along with being used as the registry for the applications.

Reviewer: I think it's important to clarify that the image registry can only be used for user application images and cannot be used for the OpenShift release or OLM operators images, see https://issues.redhat.com/browse/RFE-2488

Author: I have addressed this comment in the current preview.

// Module included in the following assemblies:
//
// scalability_and_performance/ztp-deploying-disconnected.adoc

:_module-type: CONCEPT
[id="ztp-add-local-reg-for-sno-duprofile_{context}"]
= Configuring the Image Registry Operator for local caching of images

{product-title} manages image caching with a local registry. A Single Node OpenShift (SNO) Distributed Unit (DU) is subject to bandwidth restrictions when communicating with the centralized registry. This can lead to long image download times.

This is unavoidable for the initial deployment. Over time, there is a risk that `crio` will wipe `/var/lib/containers/storage` in the case of a disorderly shutdown.

> To address this, you need a local image registry on the SNO. The cluster images need to be  mirrored to this local image registry along with being used as the registry for the applications. This is useful in Edge scenarios where an SNO is deployed on the Far Edge. The SNO cannot access the internet, just its own resources and eventually the internal network.

Reviewer: I think we should remove `The cluster images need to be mirrored to this local image registry`  since it's in contradiction with the NOTE section.

Author: Done.

[NOTE]
====

> The local image registry can only be used for user application images and cannot be used for the {product-title} or 	Operator Lifecycle Manager operator images. For details on working with these, see the information on the Topology Aware Lifecycle Manager (TAWLM) in Additional resources.

Reviewer: Topology Aware Lifecycle Manager (TAWLM) - we usually refer to it as TALM, not TAWLM

Author: Done.

====

> You can use the `SiteConfig` YAML file to generate the `MachineConfig` CR used for disk partitioning. Then use `PolicyGenTemplate` to apply to create the PV and PVC and patch `imageregistry` configuration.

Reviewer 2: Can we be a little more explicit here? Start with "This involves 2 steps: step1: Siteconfig... step 2: PGT"

Author: Done.

// Module included in the following assemblies:
//
// scalability_and_performance/ztp-advanced-policy-config.adoc

:_module-type: CONCEPT
[id="ztp-add-local-reg-for-sno-duprofile_{context}"]
= Configuring the Image Registry Operator for local caching of images

> {product-title} manages image caching with a local registry. A remote application such as {sno} is subject to bandwidth restrictions when communicating with the centralized registry, which might result in long image download times.

Reviewer 3: The term remote application is not quite correct here. We deploy applications in OpenShift. What about:

```suggestion
{product-title} manages image caching using a local registry. In edge computing use cases, clusters are often subject to bandwidth restrictions when communicating with centralized image registries, which might result in long image download times.
```

Author: Done.

> {product-title} manages image caching with a local registry. A Single Node OpenShift (SNO) Distributed Unit (DU) is subject to bandwidth restrictions when communicating with the centralized registry. This can lead to long image download times.

Reviewer 2: I guess this is required `{product-title}`  (not a typo i assume?)

Author: Variable should resolve to product name. It works fine in my preview.

> {product-title} manages image caching with a local registry. A Single Node OpenShift (SNO) Distributed Unit (DU) is subject to bandwidth restrictions when communicating with the centralized registry. This can lead to long image download times.

Reviewer 4: The third sentence feels a bit awkward on it's own. How about: 

"A Single Node OpenShift (SNO) Distributed Unit (DU) is subject to bandwidth restrictions when communicating with the centralized registry, which might result in long image download times."

> {product-title} manages image caching with a local registry. A Single Node OpenShift (SNO) Distributed Unit (DU) is subject to bandwidth restrictions when communicating with the centralized registry. A Single Node OpenShift (SNO) Distributed Unit (DU) is subject to bandwidth restrictions when communicating with the centralized registry, which might result in long image download times.

Reviewer 3: Use `{sno}` variable to refer to Single-node Openshift. See https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/term_glossary.adoc#single-node-openshift for more info.

Author: Done.

> {product-title} manages image caching with a local registry. A Single Node OpenShift (SNO) Distributed Unit (DU) is subject to bandwidth restrictions when communicating with the centralized registry. A Single Node OpenShift (SNO) Distributed Unit (DU) is subject to bandwidth restrictions when communicating with the centralized registry, which might result in long image download times.

Reviewer 3: Since deploying vDU is but one use case for ZTP, suggest reworking this intro into a more generalized form without mentioning vDU.

Author: Using "remote application".

> This is unavoidable for the initial deployment. Over time, there is a risk that `crio` will wipe `/var/lib/containers/storage` in the case of a disorderly shutdown.

Reviewer 4: As the start of a new paragraph, let's define what "this" is: 

"Long download times are unavoidable during initial deployment."

Similarly, avoid using "wipe" - let's stick with industry standard terms, like "erase". Disorderly is fine, but in Linux we typically refer to it as an "unexpected shutdown": 
"Over time, there is a risk that `crio` wil erase the `/var/lib/containers/storage` (directory, file?) in case of an unexpected shutdown."

> Long download times are unavoidable during initial deployment. Over time, there is a risk that `crio` will erase `/var/lib/containers/storage` directory in the case of an unexpected shutdown.

Reviewer 3: ```suggestion
Long download times are unavoidable during initial deployment. Over time, there is a risk that CRI-O will erase the `/var/lib/containers/storage` directory in the case of an unexpected shutdown.
```

Author: Done.

> To address this, you need a local image registry on the SNO. This is useful in Edge scenarios where an SNO is deployed on the Far Edge. The SNO cannot access the internet, just its own resources and eventually the internal network.

Reviewer 4: To address what? What is useful in Edge scenarios? Define what "this" is, because now I'm not sure if we're still talking about long download times or the risk or storage erasure.

"The SNO can only access its own resources and internal network."

Author: We are referring to long image download times here.

> To address long image download times, you need a local image registry on {sno}. This is useful in Edge scenarios where {sno} is deployed on the Far Edge. {sno} cannot access the internet, just its own resources and eventually the internal network.

Reviewer 3: I can see whaty you are getting at, but this sentence is a bit confusing. "{sno} cannot access the internet, just its own resources and eventually the internal network." What about:

```suggestion
To address long image download times, you can create a local image registry on the remote managed cluster using GitOps ZTP. This is useful in Edge computing scenarios where clusters are deployed on the far edge of the network.
```

Author: Done.

> To address long image download times, you need a local image registry on the SNO. This is useful in Edge scenarios where an SNO is deployed on the Far Edge. The SNO cannot access the internet, just its own resources and eventually the internal network.

Reviewer 3: Do not use SNO as an acronym.

Author: {sno} throughout.

[NOTE]
====

> The local image registry can only be used for user application images and cannot be used for the {product-title} or 	Operator Lifecycle Manager operator images. For details on working with these, see the information on the Topology Aware Lifecycle Manager (TALM) in Additional resources.

Reviewer 4: There is a huge space (tab?) between "or Operator" 

"For details on working with these images,..."

Use "information about..."

====

> The first phase of setting up a local image registry is configuring disk partitioning with `SiteConfig`. You can use the `SiteConfig` YAML file to generate the `MachineConfig` CR used for disk partitioning.

Reviewer 3: ```suggestion
The first phase of setting up a local image registry is configuring disk partitioning using a `SiteConfig` CR. You can use the `SiteConfig` CR to generate the `MachineConfig` CR used for disk partitioning.
```

Author: Done.

> The next phase is to configure the image registry using `PolicyGenTemplate` (PGT) CRs. You can use `PolicyGenTemplate` to create Persistent Volumes (PV) and Persistent Volume Claim (PVC) and patch `imageregistry` configuration.

Reviewer 3: ```suggestion
The next phase is to configure the image registry using `PolicyGenTemplate` CRs. The ZTP pipeline uses `PolicyGenTemplate` CRs to create Persistent Volumes (PV) and Persistent Volume Claim (PVC) CRs and patch the `imageregistry` configuration.
```
